tests/contracts/wallet_2.scilla

```diff
@@ -0,0 +1,392 @@
+    e = mk_error_event err;
+    event e
+  | True =>
+    tc <- transactionCount;
+    zero = Uint128 0;
+    amount_is_zero = builtin eq amount zero;
+    match amount_is_zero with
+    | True =>
+      (* Illegal transaction *)
+      err = InvalidAmount;
+      e = mk_error_event err;
+      event e
+    | False =>
+      (* Create new transaction *)
+      transaction = Trans recipient amount tag;
+      (* Add transaction to outstanding list of transactions *)
+      transactions[tc] := transaction; 
+      (* Sender implicitly signs *)
+      signatures[tc][_sender] := t;
+      (* Increment transaction counter *)
+      tc_new = builtin add tc transaction_inc;
+      (* Update transaction count *)
+      transactionCount := tc_new;
+      (* Create event with transaction Id *)
+      e = mk_transaction_added_event tc;
+      event e
+    end
+  end
+end
+
+(* Sign off on an existing transaction *)
+transition SignTransaction (transactionId : Uint32)
+  (* Only the owner is allowed to sign off transactions *)
+  sender_is_owner <- exists owners[_sender];
+  match sender_is_owner with
+  | False =>
+    err = NonOwnerCannotSign;
+    e = mk_error_event err;
+    event e
+  | True =>
+    (* Transaction must have been submitted *)
+    transaction <- transactions[transactionId];
+    match transaction with
+    | None =>
+      err = UnknownTransactionId;
+      e = mk_error_event err;
+      event e
+    | Some (Trans recipient amount _) =>
+      (* Sender must not have signed already *)
+      sender_has_signed <- exists signatures[transactionId][_sender];
+      match sender_has_signed with
+      | True =>
+        err = AlreadySigned;
+        e = mk_error_event err;
+        event e
+      | False =>
+        (* Signature is valid. Add to collected signatures *)
+        signatures[transactionId][_sender] := t;
+        e = mk_transaction_signed_event;
+        event e
+      end
+    end
+  end
+end
+
+(* Execute signed-off transaction *)
+transition ExecuteTransaction (transactionId : Uint32)
+  transaction_opt <- transactions[transactionId];
+  match transaction_opt with
+  | None =>
+    (* Transaction was not found. *)
+    err = UnknownTransactionId;
+    e = mk_error_event err;
+    event e
+  | Some (Trans recipient amount tag) =>
+    (* Only the recipient or an owner can execute the transaction *)
+    recipient_is_sender = builtin eq recipient _sender;
+    sender_is_owner <- exists owners[_sender];
+    sender_may_execute = orb recipient_is_sender sender_is_owner;
+    match sender_may_execute with
+    | False =>
+      err = SenderMayNotExecute;
+      e = mk_error_event err;
+      event e
+    | True =>
+      (* Check for sufficient funds  *)
+      bal <- _balance;
+      not_enough_money = builtin lt bal amount;
+      match not_enough_money with
+      | True =>
+        err = InsufficientFunds;
+        e = mk_error_event err;
+        event e
+      | False =>
+        sigs_opt <- signatures[transactionId];
```

Comment:
Just a comment: I guess this shows that having an *in-place* builtin for fetching a nested map's size could be desirable.

Reply:
@vaivaswatha and I discussed that (I don't remember who else was involved in the conversation). The problem is that the database schema doesn't support a constant-time size operation, so we need to read the entire map anyway. In other words, nothing is gained from having an in-place size operation compared to using the built-in.

The way to deal with this inefficiency is for the contract to maintain the necessary counters, which I outlined in the PR description.